lib/dslash_policy.cuh

```diff
@@ -1711,63 +1681,51 @@ namespace quda
-        break;
       case QudaDslashPolicy::QUDA_GDR_RECV_DSLASH:
-        if (!comm_gdr_blacklist())
-          result = new DslashGDRRecv<Dslash>;
-        else
-          result = new DslashBasic<Dslash>;
-        break;
+        return QudaDslashPolicy::QUDA_DSLASH;
+      case QudaDslashPolicy::QUDA_FUSED_GDR_DSLASH:
       case QudaDslashPolicy::QUDA_FUSED_GDR_RECV_DSLASH:
-        if (!comm_gdr_blacklist())
-          result = new DslashFusedGDRRecv<Dslash>;
-        else
-          result = new DslashFusedExterior<Dslash>;
-        break;
-      case QudaDslashPolicy::QUDA_ZERO_COPY_PACK_DSLASH: result = new DslashZeroCopyPack<Dslash>; break;
-      case QudaDslashPolicy::QUDA_FUSED_ZERO_COPY_PACK_DSLASH: result = new DslashFusedZeroCopyPack<Dslash>; break;
+        return QudaDslashPolicy::QUDA_FUSED_DSLASH;
       case QudaDslashPolicy::QUDA_ZERO_COPY_PACK_GDR_RECV_DSLASH:
-        if (!comm_gdr_blacklist())
-          result = new DslashZeroCopyPackGDRRecv<Dslash>;
-        else
-          result = new DslashZeroCopyPack<Dslash>;
-        break;
+        return QudaDslashPolicy::QUDA_ZERO_COPY_PACK_DSLASH;
       case QudaDslashPolicy::QUDA_FUSED_ZERO_COPY_PACK_GDR_RECV_DSLASH:
-        if (!comm_gdr_blacklist())
-          result = new DslashFusedZeroCopyPackGDRRecv<Dslash>;
-        else
-          result = new DslashFusedZeroCopyPack<Dslash>;
-        break;
-      case QudaDslashPolicy::QUDA_ZERO_COPY_DSLASH: result = new DslashZeroCopy<Dslash>; break;
-      case QudaDslashPolicy::QUDA_FUSED_ZERO_COPY_DSLASH: result = new DslashFusedZeroCopy<Dslash>; break;
-      case QudaDslashPolicy::QUDA_DSLASH_FUSED_PACK: result = new DslashFusedPack<Dslash>; break;
-      case QudaDslashPolicy::QUDA_DSLASH_FUSED_PACK_FUSED_HALO: result = new DslashFusedPackFusedHalo<Dslash>; break;
-      case QudaDslashPolicy::QUDA_DSLASH_NC: result = new DslashNC<Dslash>; break;
-      case QudaDslashPolicy::QUDA_SHMEM_UBER_PACKINTRA_DSLASH: result = new DslashShmemUberPackIntra<Dslash>; break;
-      case QudaDslashPolicy::QUDA_SHMEM_UBER_PACKFULL_DSLASH: result = new DslashShmemUberPackFull<Dslash>; break;
-      case QudaDslashPolicy::QUDA_SHMEM_PACKINTRA_DSLASH: result = new DslashShmemPackIntra<Dslash>; break;
-      case QudaDslashPolicy::QUDA_SHMEM_PACKFULL_DSLASH: result = new DslashShmemPackFull<Dslash>; break;
-      default: errorQuda("Dslash policy %d not recognized", static_cast<int>(dslashPolicy)); break;
+        return QudaDslashPolicy::QUDA_FUSED_ZERO_COPY_PACK_DSLASH;
+      default: return policy;
       }
-      return result; // default
+    }
+
+    static DslashPolicyImp<Dslash> *create(const QudaDslashPolicy &policy_)
+    {
+      // if GDR policy and blacklist enabled, create the non-GDR equivalent
+      QudaDslashPolicy policy = comm_gdr_blacklist() ? blacklist_map(policy_) : policy_;
+
+      switch (policy) {
+      case QudaDslashPolicy::QUDA_DSLASH: return new DslashBasic<Dslash>;
```

Comment:
For the `new`'s here and the `delete`'s later, should we use `std::unique_ptr`? Personally I am okay with just `new`'s and `delete`'s, but just in case that we have committed to smart pointers.

Reply:
Good suggestion.  I've switched these to `unique_ptr` (197fdde).  Small change, but good to get me in the habit of using smart pointers.